Improve Idle Timeout text (#3414)
quicwg/base-drafts: draft-ietf-quic-transport.md

## Idle Timeout {#idle-timeout}

If a max_idle_timeout is specified by either peer in transport parameters
(see {{transport-parameter-definitions}}), a connection is silently closed
and its state is discarded when it remains idle for longer than the minimum of

> the max_idle_timeouts and three times the current Probe Timeout (PTO).

Reviewer: ```suggestion
the max_idle_timeout values and three times the current Probe Timeout (PTO).
```

> the max_idle_timeout values and three times the current Probe Timeout (PTO).

Reviewer 2: ```suggestion
both max_idle_timeout values and three times the current Probe Timeout (PTO).
```

By announcing a max_idle_timeout, an endpoint commits to initiating an immediate

> close ({{immediate-close}}) if it abandons the connection prior to the effective

Reviewer 3: After removing the previous sentence, "effective value" is not defined any more.

Author: Thanks, I reverted this.  I think this was an unintentional change, but this has gone through so many iterations I'm not 100% sure.

value.

An endpoint restarts its idle timer when sending an ack-eliciting packet if no
other ack-eliciting packets have been sent since last receiving a packet.
Restarting when sending packets ensures that connections do not prematurely time

> out when initiating new activity.  An endpoint might need to send ack-eliciting

Reviewer 2: ```suggestion
Restarting this timer when sending a packet ensures that connections are not
closed after new activity is initiated.  An endpoint might need to send ack-eliciting
```

packets to avoid an idle timeout if it is unable to send application data due to
being blocked on flow control limits (see {{flow-control}}) or is application

> limited, but expecting response data.

Reviewer 4: Should "application-limited" be hyphenated?

Author: It hasn't been so far.

> limited, but expecting response data.

Reviewer 2: ```suggestion
packets to avoid an idle timeout if it is expecting response data, but does
not have or is unable to send application data.
```

An endpoint that sends packets near the end of the idle timeout period
risks having those packets discarded if its peer enters the draining state
before the packets arrive.  If a peer could time out within a Probe Timeout
(PTO; see Section 6.6 of {{QUIC-RECOVERY}}), it is advisable to send a PING or
other ack-eliciting frames to test for liveness before sending any data that
cannot be retried safely.  Note that it is likely that only applications or

> application protocols will know what information can be retried.

Reviewer 5: I don't understand this "retried safely" concept. What is an example of something unsafe to be retried?

Reviewer: You could cite RFC 8470...

Author: I've decided not to define this in this PR.  @nibanks if you'd like to do so in a different PR, feel free.

> application protocols will know what information can be retried.

Reviewer 2: This is still a bit hard to read. I'll suggest the following text instead:
```
An endpoint that sends packets close to the effective timeout risks
having them be discarded at the peer, since the peer might enter its
draining state before these packets arrive. An endpoint can send
a PING or another ack-eliciting frame to test the connection for liveness
if the peer could time out soon, such as within a PTO (see Section 6.6 of
{{QUIC-RECOVERY}}). This is especially useful if any available application
data cannot be safely retried. Note that the application determines what data
is safe to retry.
```

Author: Thanks, I accepted your suggestion.

## Idle Timeout {#idle-timeout}

If a max_idle_timeout is specified by either peer in its transport parameters
({{transport-parameter-definitions}}), the connection is silently closed
and its state is discarded when it remains idle for longer than the minimum of
both peers max_idle_timeout values and three times the current Probe Timeout
(PTO).

Each endpoint advertises a max_idle_timeout, but the effective value
at an endpoint is computed as the minimum of the two advertised values. By
announcing a max_idle_timeout, an endpoint commits to initiating an immediate
close ({{immediate-close}}) if it abandons the connection prior to the effective
value.

An endpoint restarts its idle timer when a packet from its peer is received and
processed successfully. An endpoint also restarts its idle timer when sending an
ack-eliciting packet if no other ack-eliciting packets have been sent since last
receiving and processing a packet. Restarting this timer when sending a packet
ensures that connections are not closed after new activity is initiated.

> An endpoint might need to send ack-eliciting packets to avoid an idle timeout

Reviewer: ```suggestion

An endpoint might need to send ack-eliciting packets to avoid an idle timeout
```

if it is expecting response data, but does not have or is unable to send
application data.

An endpoint that sends packets close to the effective timeout risks having
them be discarded at the peer, since the peer might enter its draining state
before these packets arrive. An endpoint can send a PING or another
ack-eliciting frame to test the connection for liveness if the peer could

> time out soon, such as within a PTO (see Section 6.6 of {{QUIC-RECOVERY}}).

Reviewer: ```suggestion
time out soon, such as within a PTO; see Section 6.6 of {{QUIC-RECOVERY}}.
```
